### PR TITLE
[Bugfix:TAGrading] Numeric grading fixes

### PR DESCRIPTION
**Pull request by Eceptonsu**
### What is the current behavior?
Users are able to enter strings and negative integers into numeric cells which will cause save errors in the back end without any feedback. At that point, a page refresh is needed. In addition, input greater than the max_clamp value for the particular component is also allowed, which will produce the same situation. 
For the comment, users can not delete text for the cell it is filled.

### What is the new behavior?
Only positive scores that are less than the max_clamp value for the component are allowed in numeric cells.
Users are able to delete text in comment cells.

### CODE REVIEW

File: site/app/templates/grading/simple/Display.twig [hunk 1]
```diff
@@ -292,15 +292,16 @@
             <td class="option-small-input">
                 <input
                     class="option-small-box cell-grade" tabindex="0"
-                    style="text-align: center; {{ component_score == 0 ? "color: var(--standard-light-medium-gray);" : "" }}
+                    style="text-align: center; width: 120px; {{ component_score == 0 ? "color: var(--standard-light-medium-gray);" : "" }}
                         {{ component_score == 0 ? "background-color: var(--alert-background-blue);" : "background-color: var(--default-white);" }}"
-                    type="text"
+                    type="number"
```
Comment (poorna2152) [inline site/app/templates/grading/simple/Display.twig:297]: Setting `type="number"` would only allow to insert integer values in to the input field. 
I think user should also be able to enter floats as well.( I assume floats also be accepted based on [this ](https://github.com/Submitty/Submitty/blob/main/site/public/js/simple-grading.js#L436)). You could also use the attribute `min=0 ` to ensure that negative values are not inserted.

File: site/public/js/simple-grading.js [hunk 2]
```diff
@@ -413,23 +413,61 @@ function setupCheckboxCells() {
...
+        var row_el = $("tr#row-" + split_id[1] + "-" + split_id[2]);
+        console.log(elem.data('origval'));
+        console.log(elem.val());
+
+        var scores = {};
+        var old_scores = {};
+        var total = 0;
+
+        var exceed_max = false;
+        const numbers = /^[0-9]+$/;
```
Comment (poorna2152) [inline site/public/js/simple-grading.js:426]: Regex only works for positive integers

File: site/public/js/simple-grading.js [hunk 3]
```diff
@@ -413,23 +413,61 @@ function setupCheckboxCells() {
...
+        var split_id = elem.attr("id").split("-");
+        var row_el = $("tr#row-" + split_id[1] + "-" + split_id[2]);
+        console.log(elem.data('origval'));
+        console.log(elem.val());
+
+        var scores = {};
+        var old_scores = {};
+        var total = 0;
+
+        var exceed_max = false;
```
Comment (poorna2152) [inline site/public/js/simple-grading.js:425]: Is this variable used.

File: site/public/js/simple-grading.js [hunk 4]
```diff
@@ -413,23 +413,55 @@ function setupCheckboxCells() {
...
+        const numbers = /^[0-9]*\.?[0-9]*$/;
+
+        if(this.tagName.toLowerCase() === 'input') {
+            // Empty input is ok for comment but not numeric cells
+            if(!this.value) {
+                this.value = 0;
+            }
+            else if(!this.value.match(numbers)) {
+                alert('Score should be a positive number');
+                this.value = 0;
```
Comment (poorna2152) [inline site/public/js/simple-grading.js:433]: return the function from here.

File: site/public/js/simple-grading.js [hunk 5]
```diff
@@ -413,23 +413,55 @@ function setupCheckboxCells() {
...
+        var old_scores = {};
+        var total = 0;
+
+        let value = this.value;
+        const numbers = /^[0-9]*\.?[0-9]*$/;
+
+        if(this.tagName.toLowerCase() === 'input') {
+            // Empty input is ok for comment but not numeric cells
+            if(!this.value) {
+                this.value = 0;
```
[the remark this answers is not in the document]
Comment (Eceptonsu) [reply site/public/js/simple-grading.js:429]: For empty input, I think it makes sense to change from an empty string to 0, and the value needs to be saved so I did not make the function returns here.

File: site/public/js/simple-grading.js (hunk 4 above)
Comment (Eceptonsu) [reply site/public/js/simple-grading.js:433]: Same as above.

File: site/public/js/simple-grading.js [hunk 6]
```diff
@@ -413,23 +413,55 @@ function setupCheckboxCells() {
...
+            }
+            else if(!this.value.match(numbers)) {
+                alert('Score should be a positive number');
+                this.value = 0;
+            }
+            // Input greater than the max_clamp for the component is not allowed
+            else {
+                submitAJAX(
+                    buildCourseUrl(['gradeable', row_el.data('gradeable'), 'maxclamp']),
```
[the remark this answers is not in the document]
Comment (Eceptonsu) [reply site/public/js/simple-grading.js:438]: Based on how the backend is written, I think at least one check is needed for both requests. But I have made some changes to make the code more concise.

**Review by cjreed121 — CHANGES_REQUESTED** [review]
Seems to work but I think it could be handled more efficiently. Rather than sending a request to check if a value is valid, just store the max clamp values in the page and have JS check them before making the request to save. The save PHP code must also perform the check before attempting to save to the database.

File: site/app/templates/grading/simple/Display.twig [hunk 7]
```diff
@@ -292,15 +292,16 @@
...
                     id="cell-{{ section_id }}-{{ index }}-{{ loop.index0 }}"
                     data-id="{{ component.getId() }}"
                     data-origval="{{ component_score }}"
                     value="{{ component_score }}"
```
Comment (cjreed121) [inline site/app/templates/grading/simple/Display.twig:301]: Why not store the max clamp in a place like this? I think the split will work but I think putting data attributes feels cleaner.

File: site/public/js/simple-grading.js [hunk 8]
```diff
@@ -413,23 +413,42 @@ function setupCheckboxCells() {
...
+            }
+            else if(!this.value.match(numbers)) {
+                alert('Score should be a positive number');
+                this.value = 0;
+            }
+            // Input greater than the max_clamp for the component is not allowed
+            else {
+                let max_clamps = localStorage.getItem('max_clamp').split('-');
+                if(max_clamps[split_id[3]] && max_clamps[split_id[3]] < this.value) {
+                    alert('Score should be less than the max clamp value: ' + max_clamps[split_id[3]]);
```
Comment (cjreed121) [inline site/public/js/simple-grading.js:439]: This should probably be "less than or equal to" rather than just "less than"

**Review by cjreed121 — CHANGES_REQUESTED** [review]
I also noticed an issue. If I had a max clamp of 5 and put in a number greater than 5 but less than 10 it would do an alert. But if I did 10 there was no alert but there was a message that went in the JS console.

**Comment by Eceptonsu** [comment]
> 

Thanks for pointing out, changes have been made.

File: site/app/templates/grading/simple/Display.twig [hunk 9]
```diff
@@ -122,7 +122,7 @@
     {% endif %}
 
     {# This is a data table #}
-    <table id="data-table" data-current-grader="{{ user_id }}" class="table table-striped table-bordered persist-area" style="table-layout: fixed;">
+    <table id="data-table" data-gradeable="{{ gradeable.getId() }}" data-current-grader="{{ user_id }}" class="table table-striped table-bordered persist-area" style="table-layout: fixed;">
```
Comment (cjreed121) [inline site/app/templates/grading/simple/Display.twig:125]: Why add data-gradeable?

**Review by cjreed121 — CHANGES_REQUESTED** [review]
After this last requested change I think this should be good to go.

File: site/app/templates/grading/simple/Display.twig (hunk 9 above)
Comment (Eceptonsu) [reply site/app/templates/grading/simple/Display.twig]: Forgot to delete this change from previous commits, should be good now.

**Review by cjreed121 — APPROVED** [review]

**Pull request merged**